lib/srv/authhandlers.go

```diff
@@ -487,47 +527,107 @@ func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*s
+	// the git forwarding component currently only supports an authorization model that makes sense
+	// for local identities. reject all non-local identities explicitly.
+	if h.c.Component == teleport.ComponentForwardingGit && clusterName.GetClusterName() != ca.GetClusterName() {
+		log.ErrorContext(ctx, "cross-cluster git forwarding is not supported", "local_cluster", clusterName.GetClusterName(), "remote_cluster", ca.GetClusterName())
+		recordFailedLogin(trace.AccessDenied("cross-cluster git forwarding is not supported"))
+		return nil, trace.AccessDenied("cross-cluster git forwarding is not supported")
+	}
+
+	var accessPermit *decisionpb.SSHAccessPermit
+	var gitForwardingPermit *GitForwardingPermit
+	var proxyPermit *proxyingPermit
+	var diagnosticTracing bool
+
+	switch h.c.Component {
+	case teleport.ComponentForwardingGit:
+		gitForwardingPermit, err = h.evaluateGitForwarding(ident, ca, clusterName.GetClusterName(), h.c.TargetServer)
+	case teleport.ComponentProxy:
+		proxyPermit, err = h.evaluateProxying(ident, ca, clusterName.GetClusterName())
+	case teleport.ComponentForwardingNode:
+		diagnosticTracing = true
 		if h.c.TargetServer != nil && h.c.TargetServer.IsOpenSSHNode() {
-			err = h.canLoginWithRBAC(ident, ca, clusterName.GetClusterName(), h.c.TargetServer, conn.User())
+			accessPermit, err = h.evaluateSSHAccess(ident, ca, clusterName.GetClusterName(), h.c.TargetServer, conn.User())
+		} else {
+			proxyPermit, err = h.evaluateProxying(ident, ca, clusterName.GetClusterName())
 		}
-	} else {
-		// the SSH server is a Teleport node, preform an RBAC check now
-		err = h.canLoginWithRBAC(ident, ca, clusterName.GetClusterName(), h.c.Server.GetInfo(), conn.User())
+	case teleport.ComponentNode:
+		diagnosticTracing = true
+		accessPermit, err = h.evaluateSSHAccess(ident, ca, clusterName.GetClusterName(), h.c.Server.GetInfo(), conn.User())
+	default:
+		return nil, trace.BadParameter("cannot determine appropriate authorization checks for unknown component %q (this is a bug)", h.c.Component)
 	}
+
 	if err != nil {
-		log.ErrorContext(ctx, "Permission denied", "error", err)
+		log.ErrorContext(ctx, "permission denied",
+			"error", err,
+			"local_addr", logutils.StringerAttr(conn.LocalAddr()),
+			"remote_addr", logutils.StringerAttr(conn.RemoteAddr()),
+			"key", key.Type(),
+			"fingerprint", sshutils.Fingerprint(key),
+			"user", cert.KeyId,
+		)
+
 		recordFailedLogin(err)
+
+		if h.c.OnRBACFailure != nil {
+			h.c.OnRBACFailure(conn, ident, err)
+		}
+
 		return nil, trace.Wrap(err)
 	}
 
-	if err := h.maybeAppendDiagnosticTrace(ctx, ident.ConnectionDiagnosticID,
-		types.ConnectionDiagnosticTrace_RBAC_NODE,
-		"You have access to the Node.",
-		nil,
-	); err != nil {
-		return nil, trace.Wrap(err)
+	if accessPermit != nil {
+		encodedPermit, err := utils.FastMarshal(accessPermit)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		permissions.Extensions[utils.ExtIntSSHAccessPermit] = string(encodedPermit)
```

Comment:
Why do we need to extend the `permissions` map from the SSH cert? Aren't we turning the whole SSH cert into a permit?